src/backend/distributed/executor/multi_executor.c

```diff
@@ -24,273 +24,384 @@
+	StringInfo jobDirectoryName = MasterJobDirectoryName(workerJob->jobId);
+	CreateDirectory(jobDirectoryName);
 
-	if (eflags & EXEC_FLAG_CITUS_ROUTER_EXECUTOR)
+	ResourceOwnerEnlargeJobDirectories(CurrentResourceOwner);
+	ResourceOwnerRememberJobDirectory(CurrentResourceOwner, workerJob->jobId);
+}
+
+
+/*
+ * Load data collected by real-time or task-tracker executors into the tuplestore
+ * of CitusScanState. For that first create a tuplestore, and then copy the
+ * files one-by-one.
+ *
+ * Note that in the long term it'd be a lot better if Multi*Execute() directly
+ * filled the tuplestores, but that's a fair bit of work.
+ */
+static void
+LoadTuplesIntoTupleStore(CitusScanState *scanState, Job *workerJob)
+{
+	CustomScanState customScanState = scanState->customScanState;
+	List *workerTaskList = workerJob->taskList;
+	EState *executorState = NULL;
+	MemoryContext executorTupleContext = NULL;
+	ExprContext *executorExpressionContext = NULL;
+	TupleDesc tupleDescriptor = NULL;
+	Relation fauxRelation = NULL;
+	ListCell *workerTaskCell = NULL;
+	uint32 columnCount = 0;
+	Datum *columnValues = NULL;
+	bool *columnNulls = NULL;
+
+	executorState = scanState->customScanState.ss.ps.state;
+	executorTupleContext = GetPerTupleMemoryContext(executorState);
+	executorExpressionContext = GetPerTupleExprContext(executorState);
+
+	tupleDescriptor = customScanState.ss.ps.ps_ResultTupleSlot->tts_tupleDescriptor;
+	fauxRelation = FauxRelation(tupleDescriptor);
+
+	columnCount = tupleDescriptor->natts;
+	columnValues = palloc0(columnCount * sizeof(Datum));
+	columnNulls = palloc0(columnCount * sizeof(bool));
+
+	Assert(scanState->tuplestorestate == NULL);
+	scanState->tuplestorestate = tuplestore_begin_heap(true, false, work_mem);
+
+	foreach(workerTaskCell, workerTaskList)
 	{
-		/* drop into the router executor */
-		RouterExecutorEnd(queryDesc);
+		Task *workerTask = (Task *) lfirst(workerTaskCell);
+		StringInfo jobDirectoryName = MasterJobDirectoryName(workerTask->jobId);
+		StringInfo taskFilename = TaskFilename(jobDirectoryName, workerTask->taskId);
+		List *copyOptions = NIL;
+		CopyState copyState = NULL;
+
+		if (BinaryMasterCopyFormat)
+		{
+			DefElem *copyOption = makeDefElem("format", (Node *) makeString("binary"));
+			copyOptions = lappend(copyOptions, copyOption);
+		}
+
+		copyState = BeginCopyFrom(fauxRelation, taskFilename->data, false, NULL,
+								  copyOptions);
+
+		while (true)
+		{
+			MemoryContext oldContext = NULL;
+			bool nextRowFound = false;
+
+			ResetPerTupleExprContext(executorState);
+			oldContext = MemoryContextSwitchTo(executorTupleContext);
+
+			nextRowFound = NextCopyFrom(copyState, executorExpressionContext,
+										columnValues, columnNulls, NULL);
+			if (!nextRowFound)
+			{
+				MemoryContextSwitchTo(oldContext);
+				break;
+			}
+
+			tuplestore_putvalues(scanState->tuplestorestate, tupleDescriptor,
+								 columnValues, columnNulls);
+			MemoryContextSwitchTo(oldContext);
+		}
```

Comment:
Needs an ```EndCopyFrom(copyState);``` otherwise we're (temporarily) leaking filehandles.  Noticed when playing around with a high shard count ;)